_posts/2019-03-30-build-js-library-with-webpack-ringcentral-c2d.md

```diff
@@ -0,0 +1,122 @@
+---
+layout: post
+title:  "How to build a JS library with Webpack -- RingCentral C2D library"
+date:   2019-03-30 10:00:00 +0800
+categories: blog
+author: Embbnux Ji
+---
+
+![clicktodial](https://user-images.githubusercontent.com/7036536/51652788-d2627200-1fcb-11e9-8ba3-9e50baeaf8a6.png)
+
+Rencently, we publish RingCentral C2D library to help developers integrate RingCentral Click To Dial in browser. This library is built with Webpack 4 and Typescript. This article will share how to build a JS library which support CDN and NPM ways by introducing how we build RingCentral C2D.
+
+## Introduction about RingCentral C2D
+
+RingCentral C2D aims to help developers to get phone numbers in web page and pass it to our RingCentral integration app.
+
+#### Project Links:
+
+* Github Repo: [https://github.com/ringcentral/ringcentral-c2d](https://github.com/ringcentral/ringcentral-c2d)
+* Demo Page: [https://ringcentral.github.io/ringcentral-c2d/](https://ringcentral.github.io/ringcentral-c2d/)
+
+#### Built with:
+
+* Webpack 4
+* Typescript
+
+## Using webpack 4
+
+### Output option
+
+Webpack 4 provides very easy way to build JS library with several options:
+
+```js
+// webpack config object
+config.ouput = {
+  filename: 'index.js',
+  path: path.resolve(__dirname, 'build'),
+  library: 'RingCentralC2DInject',
+  libraryTarget: 'umd',
+  libraryExport: 'default'
+}
+```
+
+In output config, `filename` and `path` define where to output your library. `config.library` and `config.libraryTarget` property tells webpack you are building a library not an app. And with `umd`, this library will support to used with CommonJS, AMD and as global variable. When this library is used with CDN script, it will export `RingCentralC2DInject` as global variable.
+
+With `libraryExport: 'default'`, webpack will add `default` in module exports. So you can use as `import RingCentralC2D from 'ringcentral-c2d'`.
+
+### TypeScript loader
+
+Because our library source codes were written with TypeScript, we need to make webpack support it. Just add TS loader in `config.module.rules`:
+
+```js
+config.module = {
+  rules: [
+    {
+      test: /\.ts?$/,
+      use: 'ts-loader',
+      exclude: /node_modules/
+    }
+  ]
+}
+```
+
+After that, webpack support to compile and bundle TS codes.
+
+### Externals option
+
+Our library is based on `libphonenumber-js` library. But we should not bundle it into our library. Developers should allow to import it themselves. We can use `externals` option:
+
+```
+config.externals = {
+  'libphonenumber-js': {
+    commonjs: 'libphonenumber-js',
+    commonjs2: 'libphonenumber-js',
+    amd: 'libphonenumber-js',
+    root: 'libphonenumber'
+  }
+}
+```
+
+We allow developers to import `libphonenumber-js` with CommonJS, AMD and global variable.
+
+So developers can use our library with `libphonenumber-js` CDN version:
```

Comment:
So developers can use `the CDN hosted version of` `libphonenumber-js` with our library: